src/main/java/dev/morling/onebrc/CalculateAverage_yavuztas.java

```diff
@@ -96,111 +78,175 @@ public int hashCode() {
-            return new String(bytes, 0, this.length, StandardCharsets.UTF_8);
+            this.buffer.get(this.start, bytes);
+            return new String(bytes, StandardCharsets.UTF_8);
+        }
+
+        public void collect(int temp) {
+            this.min = Math.min(this.min, temp);
+            this.max = Math.max(this.max, temp);
+            this.sum += temp;
+            this.count++;
+        }
+
+        public void merge(Record other) {
+            this.min = Math.min(this.min, other.min);
+            this.max = Math.max(this.max, other.max);
+            this.sum += other.sum;
+            this.count += other.count;
+        }
+
+        public String measurements() {
+            final StringBuilder sb = new StringBuilder(14);
+            sb.append(this.min / 10.0); // convert to double while generating the string output
+            sb.append("/");
+            sb.append(round((this.sum / 10.0) / this.count));
+            sb.append("/");
+            sb.append(this.max / 10.0);
+            return sb.toString();
         }
     }
 
+    // Inspired by @spullara - customized hashmap on purpose
+    // The main difference is this map holds only one array instead of two
+    static class KeyBufferMap {
+
+        static final int SIZE = 1 << 15; // 32k - bigger bucket size less collisions
+        static final int BITMASK = SIZE - 1;
+        Record[] keys = new Record[SIZE];
+
+        private int hashBucket(int hash) {
+            hash = hash ^ (hash >>> 16); // naive bit spreading but greatly decreases collision
+            return hash & BITMASK; // fast modulo, to find bucket
+        }
+
+        void putAndCollect(Record key, int temp) {
+            int bucket = hashBucket(key.hash);
+            Record existing = this.keys[bucket];
+            if (existing == null) {
+                this.keys[bucket] = key;
+                key.collect(temp);
+                return;
+            }
+
+            if (!existing.equals(key)) {
+                // collision, linear probing to find a slot
+                while ((existing = this.keys[++bucket]) != null && !existing.equals(key)) {
+                    // no-op
+                }
+                if (existing == null) {
+                    this.keys[bucket] = key;
+                    key.collect(temp);
+                    return;
+                }
+                existing.collect(temp);
+            }
+            else {
+                existing.collect(temp);
+            }
+        }
+
+        void putOrMerge(Record key) {
+            int bucket = hashBucket(key.hash);
+            Record existing = this.keys[bucket];
+            if (existing == null) {
+                this.keys[bucket] = key;
+                return;
+            }
+
+            if (!existing.equals(key)) {
+                // collision, linear probing to find a slot
+                while ((existing = this.keys[++bucket]) != null && !existing.equals(key)) {
```

Comment:
Since you don't use `% keys.length` here, an unlucky case where the collision occurs near the end of the array will result in `IndexOutOfBoundsException`.

Reply:
Oh yes, absolutely. Thanks, @mtopolnik, I will fix it.

Reply:
fixed